File: server/Models/Entities/ResearchGroup.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace server.Models.Entities
{
    public class ResearchGroup
    {
        [Key]
        [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
        public Guid Id { get; set; }
        [MaxLength(100)]
        public string GroupName { get; set; } = string.Empty;

        public string ResearchTitle { get; set; } = string.Empty;
        [MaxLength(128)]
        public string? AdviserId { get; set; }
        [MaxLength(128)]
        public string LeaderId { get; set; } = string.Empty;


        public DateTime CreatedAt { get; set; }

        public DateTime UpdatedAt { get; set; }
        public virtual ICollection<Student> Students { get; set; } = [];
        public Faculty?  Adviser { get; set; }

        // Leader = the student referenced by LeaderId. Needed to resolve the
        // group's institute for the schedule calendar card colors.
        public Student? Leader { get; set; }

        public ICollection<Schedule> Schedules { get; set; } = [];

    }
}

File: packages/types/src/research-group.ts
export interface CreateResearchGroupDto {
  groupName: string;
  researchTitle: string;
  adviserId?: string;
  leaderId?: string;
}

export interface UpdateResearchGroupDto {
  groupName?: string;
  researchTitle?: string;
  adviserId?: string;
  leaderId?: string;
}

export interface ResearchGroupResponseDto {
  id: string;
  groupName: string;
  researchTitle: string;
  adviserId: string;
  leaderId: string;
  createdAt: string;
  updatedAt: string;
  institute: string;
  members: ResearchGroupMemberDto[];
}

export interface ResearchGroupMemberDto {
  id: string;
  studentNumber: string;
  name: string;
  position: string;
  program: string;
}

export type ResearchGroupResponseListDto = ResearchGroupResponseDto[];
